### PR TITLE
feat(worker): Google Analytics 4 experiment

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

**Comment by ry** [comment]
Can you move the ga4 stuff into deno.land/x/ga4 so we can share it between multiple projects?

How will this effect the current "deno for bots" GA account?

**Comment by piscisaureus** [comment]
> Can you move the ga4 stuff into deno.land/x/ga4 so we can share it between multiple projects?

I will eventually - but there are still a number of kinks to iron out and I also this abstraction isn't right yet.
You can see the status quo here: https://analytics.google.com/analytics/web/#/p311458244/realtime/overview

> How will this effect the current "deno for bots" GA account?

For now it reports to both. The current GA account is unaffected.

File: worker/handler.ts [hunk 1]
```diff
@@ -78,6 +80,111 @@ const ga: Reporter = async (...args) => {
...
+      return;
+    }
+
+    // Filter out traffic generated by uptime checks (Checkly and GCP).
+    if (
+      botName === "Checkly" ||
+      botName === "GoogleStackdriverMonitoring-UptimeChecks"
+    ) {
+      return;
+    }
```
Comment (ry) [inline worker/handler.ts:112]: Maybe these should be filtered out elsewhere (i.e. during analysis)

If we rely on this, then a new check is added, or the UA changes or checkly, then the data will be thrown off and we won't really understand why.

File: worker/handler.ts [hunk 2]
```diff
@@ -78,6 +80,111 @@ const ga: Reporter = async (...args) => {
...
+
+    const bot = isBot(request);
+    const botName = bot ? getBotName(request) : null;
+
+    // If the visitor is using a web browser, only create events when we serve
+    // a top level documents or download; skip assets like css, images, fonts.
+    if (!bot && !isNonAssetRequest(request, response) && error == null) {
+      return;
+    }
```
Comment (ry) [inline worker/handler.ts:104]: Here too. It seems less error prone if we are non relying on these hardcoded heuristics.

File: completions_v2.ts [hunk 3]
```diff
@@ -226,3 +226,26 @@ export const routes = {
...
+      }
+    }
+  }
+  return Promise.resolve(
+    new Response(null, {
+      status: 404,
+      statusText: "NotFound",
+    }),
+  );
+}
```
Comment (crowlKats) [inline completions_v2.ts:251]: this is not necessary anymore

**Pull request merged**